File: runtime-rust/src/resolver.rs
use std::collections::HashMap;
use std::net::ToSocketAddrs;
use std::sync::{Arc, Mutex};

pub trait Resolver: Send + Sync + 'static {
    fn resolve(&self, service_name: &str) -> Result<Vec<String>, String>;
}

#[derive(Clone)]
pub struct StaticResolver {
    targets: Arc<Mutex<HashMap<String, Vec<String>>>>,
}

impl Default for StaticResolver {
    fn default() -> Self {
        Self::new()
    }
}

impl StaticResolver {
    pub fn new() -> Self {
        StaticResolver {
            targets: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    pub fn register(&self, service_name: &str, addresses: Vec<String>) {
        let mut guard = self.targets.lock().unwrap();
        guard.insert(service_name.to_string(), addresses);
    }
}

impl Resolver for StaticResolver {
    fn resolve(&self, service_name: &str) -> Result<Vec<String>, String> {
        let guard = self.targets.lock().unwrap();
        guard
            .get(service_name)
            .cloned()
            .ok_or_else(|| format!("service {} not found in resolver targets", service_name))
    }
}

pub struct DnsResolver;

impl Default for DnsResolver {
    fn default() -> Self {
        Self::new()
    }
}

impl DnsResolver {
    pub fn new() -> Self {
        DnsResolver
    }
}

impl Resolver for DnsResolver {
    fn resolve(&self, service_name: &str) -> Result<Vec<String>, String> {
        let lookup_target = if service_name.contains(':') {
            service_name.to_string()
        } else {
            format!("{}:8080", service_name)
        };

        match lookup_target.to_socket_addrs() {
            Ok(addrs) => {
                let ips: Vec<String> = addrs.map(|addr| addr.to_string()).collect();
                if ips.is_empty() {
                    Err(format!("No IP addresses resolved for {}", service_name))
                } else {
                    Ok(ips)
                }
            }
            Err(e) => Err(e.to_string()),
        }
    }
}

File: runtimes/python/helix_rt/resolver.py
import socket

class Resolver:
    def resolve(self, service_name: str) -> list[str]:
        raise NotImplementedError

class StaticResolver(Resolver):
    def __init__(self):
        self.targets = {}

    def register(self, service_name: str, addresses: list[str]) -> None:
        self.targets[service_name] = addresses

    def resolve(self, service_name: str) -> list[str]:
        if service_name not in self.targets:
            raise ValueError(f"service {service_name} not found in resolver targets")
        return self.targets[service_name]

class DNSResolver(Resolver):
    def resolve(self, service_name: str) -> list[str]:
        if ":" in service_name:
            host, port = service_name.split(":", 1)
            port = int(port)
        else:
            host = service_name
            port = 8080

        try:
            infos = socket.getaddrinfo(host, port, proto=socket.IPPROTO_TCP)
            ips = list(set([f"{info[4][0]}:{port}" for info in infos]))
            if not ips:
                raise ValueError(f"No IP addresses resolved for {service_name}")
            return ips
        except Exception as e:
            raise ValueError(f"Failed to resolve DNS target {service_name}: {e}")
